src/ann_solo/spectrum.py

```diff
@@ -214,6 +215,57 @@ def spectrum_to_vector(spectrum: MsmsSpectrum, min_mz: float, max_mz: float,
     return vector
 
 
+def spectrum_to_vector(spectrum: MsmsSpectrum, transformation: ss.csr_matrix,
+                       min_mz: float, max_mz: float, bin_size: float, dim: int,
+                       norm: bool) -> np.ndarray:
+    """
+    Convert a single spectrum to a dense NumPy vector.
+
+    Peaks are first discretized to mass bins of width `bin_size` starting from
+    `min_mz`, after which they are transformed using sparse random projections.
+
+    Parameters
+    ----------
+    spectrum : MsmsSpectrum
+        The spectrum to be converted to a vector.
+    transformation : ss.csr_matrix
+        Sparse random projection transformation to convert sparse spectrum
+        vectors to low-dimensional dense vectors.
+    min_mz : float
+        The minimum m/z to include in the vector.
+    max_mz : float
+        The maximum m/z to include in the vector.
+    bin_size : float
+        The bin size in m/z used to divide the m/z range.
+    dim : int
+        The high-resolution vector dimensionality.
+    norm : bool
+        Normalize the vector to unit length or not.
+    Returns
+    -------
+    np.ndarray
+        The low-dimensional transformed spectrum vector with unit length.
+    """
+    # set the spectrum range between min and max mz
+    spectrum = spectrum.set_mz_range(min_mz, max_mz)
+    # Convert a spectrum to a binned sparse vector
+    data = np.array(spectrum.intensity, dtype=np.float32)
+    indices = np.array(
+        [math.floor((mz - min_mz) / bin_size) for mz in spectrum.mz],
+        dtype=np.int32)
+    indptr = np.array([0, len(spectrum.mz)], dtype=np.int32)
+
+    # Instantiate the sparse matrix
+    sparse_vector = ss.csr_matrix(
+        (data, indices, indptr), (1, dim), np.float32, False)
+
+    # Transform
+    transformed_vector = (sparse_vector @ transformation).toarray()
+    if norm:
+        transformed_vector /= np.linalg.norm(transformed_vector)
```

Comment:
**comment:** Maybe there could be a small performance increase by [computing the norm on the sparse vector](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.norm.html) and only afterwards converting to a dense vector?

Reply:
I ll modify the transformation to Gaussian projection, and we'll need no further conversion to dense vector. After the last dot product.